---
title: "Bap1 Glia Differential Gene Expression"
author: "Kat Beigel (beigelk@chop.edu)"
date: "`r Sys.Date()`"
output: html_document
---

# Glia KO. v. WT by cluster, using Seurat v3.1.2

```{bash, include=FALSE}

wget https://github.com/satijalab/seurat/archive/refs/tags/v3.1.2.tar.gz

```

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)

# Using renv because I need to use Seurat v3.1.2
# renv::init(bare = TRUE)

```

```{r}

# install.packages("remotes")
# install.packages("devtools")
# 
# library(remotes)
# library(devtools)
# 
# remotes::install_version("SDMTools", "1.1-221")
# install_local("Seurat_DifferentialExpression/pkg_download/v3.1.2.tar.gz")
# 
# install.packages("mgsub")
# install.packages("tidyverse")
#
# renv::snapshot()

```


```{r Library}

# Utility 
library(mgsub)
library(tidyverse)
library(xlsx)

# Analysis
library(Seurat)

```

```{r Directores}

datadir = "Seurat_DifferentialExpression/data"
resdir = "Seurat_DifferentialExpression/results"

```


```{r pressure, echo=FALSE}

Combined_Data_Glia_ClusterNames_v2 = readRDS(file = file.path(datadir, "Combined_Data_Glia_ClusterNames_v2.rds"))

Combined_Data_Glia_ClusterNames_v2 = AddMetaData(Combined_Data_Glia_ClusterNames_v2,
                                                 unname(Idents(Combined_Data_Glia_ClusterNames_v2)),
                                                 col.name = "celltype")

DimPlot(Combined_Data_Glia_ClusterNames_v2)

Idents(Combined_Data_Glia_ClusterNames_v2) = "celltype"
DimPlot(Combined_Data_Glia_ClusterNames_v2)

Idents(Combined_Data_Glia_ClusterNames_v2) = "orig.ident"
DimPlot(Combined_Data_Glia_ClusterNames_v2)

find_all_markers_res_all = FindAllMarkers(Combined_Data_Glia_ClusterNames_v2, assay = "RNA")
write.csv(find_all_markers_res_all, paste0(resdir, '/', 'Glia/', 'GliaAll', '_', 'Combined_Data_Glia_ClusterNames_v2', '.csv'))

write.xlsx(find_all_markers_res_all, file =  paste0(resdir, '/', 'Glia/', 'Glia_GeneExpression_ConditionsCompared', '.xlsx'),
           sheetName =  "All_Clusters", col.names = TRUE, row.names = TRUE, append = FALSE)

```

```{r Getting differentially expressed genes}


# Following S. Schneider's method:
# index <- 1
# for (val in sequence)
# {
#   cluster <- subset(x = CombinedData_ClusterNames, idents = val)
#   Idents(cluster) <- "orig.ident"
#   name <- names[index]
#   write.csv(FindAllMarkers(cluster, assay = "RNA"), paste("G:\\Single Cell Seq Bap1 Data\\Bap1GeneExpression\\", name, ".csv"))
#   index <- index + 1
# }

# Updating this loop to make it easier
Idents(Combined_Data_Glia_ClusterNames_v2) = "celltype"
list_of_clusters = unique(Combined_Data_Glia_ClusterNames_v2@meta.data$celltype)
# names(list_of_clusters) = unique(Combined_Data_Glia_ClusterNames_v2@meta.data$seurat_clusters)

for (item in 1:length(list_of_clusters)) {
  
  # clusternum = names(list_of_clusters)[item]
  cluster = as.character(unname(list_of_clusters[item]))

  cluster_subset <- subset(x = Combined_Data_Glia_ClusterNames_v2, idents = cluster)
  Idents(cluster_subset) <- "orig.ident"
  find_all_markers_res = FindAllMarkers(cluster_subset,
                                        assay = "RNA")

  # write.csv(find_all_markers_res, paste0(resdir, '/', 'Glia/ClusterNames/', 'Glia_GeneExpression_ConditionsCompared', '_',
  #                                        mgsub(cluster, c(' ', ', ', '/'), c('_', '_', '_')), '.csv'),
  #           quote=FALSE, row.names = TRUE)
  # 
  
  write.xlsx(find_all_markers_res, file =  paste0(resdir, '/', 'Glia/', 'Glia_GeneExpression_ConditionsCompared', '.xlsx'),
             sheetName =  mgsub(cluster, c(' ', ', ', '/', '\\?'), c('_', '_', '_', '')), col.names = TRUE, row.names = TRUE, append = TRUE)
  
}




```


```{r pressure, echo=FALSE}

Combined_Data_Glia_Cluster_v2 = readRDS(file = file.path(datadir, "Combined_Data_Glia_Cluster_v2.rds"))

Combined_Data_Glia_Cluster_v2 = AddMetaData(Combined_Data_Glia_Cluster_v2,
                                                 unname(Idents(Combined_Data_Glia_Cluster_v2)),
                                                 col.name = "celltype")

DimPlot(Combined_Data_Glia_Cluster_v2)

```

```{r Getting differentially expressed genes}
# Following S. Schneider's method:
# index <- 1
# for (val in sequence)
# {
#   cluster <- subset(x = CombinedData_ClusterNames, idents = val)
#   Idents(cluster) <- "orig.ident"
#   name <- names[index]
#   write.csv(FindAllMarkers(cluster, assay = "RNA"), paste("G:\\Single Cell Seq Bap1 Data\\Bap1GeneExpression\\", name, ".csv"))
#   index <- index + 1
# }

# Updating this loop to make it easier
list_of_clusters = unique(Combined_Data_Glia_Cluster_v2@meta.data$celltype)

for (item in 1:length(list_of_clusters)) {
  
  # clusternum = names(list_of_clusters)[item]
  cluster = as.character(unname(list_of_clusters[item]))

  cluster_subset <- subset(x = Combined_Data_Glia_Cluster_v2, idents = cluster)
  Idents(cluster_subset) <- "orig.ident"
  find_all_markers_res = FindAllMarkers(cluster_subset,
                                        assay = "RNA")

  write.csv(find_all_markers_res, paste0(resdir, '/', 'Glia/Cluster/', 'Glia', '_',
                                         mgsub(cluster, c(' ', ', ', '/'), c('_', '_', '_')), '.csv'))


  
  }

```



```{r}

sink("Seurat_Glia_DEGs_byCluster_sessionInfo.txt")
sessionInfo()
sink()

```
